pepdbagent/db_utils.py

```diff
@@ -0,0 +1,173 @@
+    submission_date: Mapped[datetime.datetime]
+    last_update_date: Mapped[datetime.datetime]
+    pep_schema: Mapped[Optional[str]]
+
+    __table_args__ = (PrimaryKeyConstraint("namespace", "name", "tag", name="id"),)
+
+
+class BaseEngine:
+    """
+    A class with base methods, that are used in several classes. e.g. fetch_one or fetch_all
+    """
+
+    def __init__(
+        self,
+        *,
+        host: str = "localhost",
+        port: int = 5432,
+        database: str = "pep-db",
+        user: str = None,
+        password: str = None,
+        drivername: str = POSTGRES_DIALECT,
+        dsn: str = None,
+        echo: bool = False,
+    ):
+        """
+        Initialize connection to the pep_db database. You can use The basic connection parameters
+        or libpq connection string.
+        :param host: database server address e.g., localhost or an IP address.
+        :param port: the port number that defaults to 5432 if it is not provided.
+        :param database: the name of the database that you want to connect.
+        :param user: the username used to authenticate.
+        :param password: password used to authenticate.
+        :param drivername: driver used in
+        :param dsn: libpq connection string using the dsn parameter
+        (e.g. 'postgresql://user_name:password@host_name:port/db_name')
+        """
+        if not dsn:
+            dsn = URL.create(
+                host=host,
+                port=port,
+                database=database,
+                username=user,
+                password=password,
+                drivername=drivername,
+            )
+
+        self._engine = create_engine(dsn, echo=echo)
+        self.create_schema(self._engine)
+        self.check_db_connection()
+
+    def create_schema(self, engine=None):
+        """
+        Create sql schema in the database.
+
+        :param engine: sqlalchemy engine [Default: None]
+        :return: None
+        """
+        if not engine:
+            engine = self._engine
+        Base.metadata.create_all(engine)
+        return None
+
+    def session_execute(self, statement: Select) -> Result:
+        """
+        Execute statement using sqlalchemy statement
+
+        :param statement: SQL query or a SQL expression that is constructed using
+            SQLAlchemy's SQL expression language
+        :return: query result represented with declarative base
+        """
+        _LOGGER.debug(f"Executing statement: {statement}")
+        with Session(self._engine) as session:
+            query_result = session.execute(statement)
+
+        return query_result
+
+    @property
+    def session(self):
+        """
+        :return: started sqlalchemy session
+        """
+        return self._start_session()
+
+    @property
+    def engine(self):
+        return self._engine
+
+    def _start_session(self):
+        session = Session(self.engine)
+        try:
+            session.execute(select(Projects)).first()
```

Comment:
Can we make an even simpler query? I like to use 
```sql
SELECT 1
```

for `postgres` to check connections. This makes it as simple as possible and schema no longer is a variable. Could we just do:

```python
self.engine.execute("SELECT 1")
```

Reply:
I am not testing connection, I am testing if schema is correct. So we can's use SELECT 1 in this case

Reply:
And actually there is an error. it should be session.execute(select(Projects).limit(1))

Reply:
Oh ok. I made a comment earlier under `def check_db_connection` too where I think `SELECT 1` could be good.